### String Functions
- upper() -- Will convert the input string into upper case
- Lower() --Will converts the input string into lower case

In [1]:
s1 = 'Python'
print(s1.upper())
print(s1.lower())

PYTHON
python


### String Boolean Function
- True or False
- islower() -- True if the string have all characters lower case
- isupper(0 -- True if the string have all characters upper case
- istitle() -- True if the string in title case
- isnumeric() -- True if the string numeric
- isalpha() -- True if the string only alphabets
- isspace() -- True if the string is space

In [4]:
s1 = "PYTHON"
print(s1.islower())
print(s1.isupper())

False
True


In [5]:
s1 = "python programming"
print(s1.istitle())

False


In [6]:
s1 ="123456"
s2 = "py123"
print(s1.isnumeric())
print(s2.isnumeric())

True
False


In [9]:
s1 ="py123456"
s2 = "python"
print(s1.isalpha())
print(s2.isalpha())

False
True


In [11]:
s1 = ' '
s2 = '1 2 3'
print(s1.isspace())
print(s2.isspace())


True
False


### String Methods
- split() -- Returns the list of strings seperated by whitspace
- replace() -- Returns the string replace by word/character
- join() -- Concatination of two strings

In [12]:
s1 = "python programming Learn Made Easy"
li = s1.split()
print(li)

['python', 'programming', 'Learn', 'Made', 'Easy']


In [15]:
s1 = "python programming Learn Made Easy"
li = s1.split('a')
print(li)

['python progr', 'mming Le', 'rn M', 'de E', 'sy']


In [17]:
s1 = "python programming Learn Made Easy"
print(s1.replace(' ','-'))

python-programming-Learn-Made-Easy


In [18]:
s1 = "python programming Learn Made Easy"
print(s1.replace('a','@'))

python progr@mming Le@rn M@de E@sy


In [21]:
s1 = "python programming"
print(" ".join(s1))

p y t h o n   p r o g r a m m i n g


### String Formatting
- Classic version
- Current version python

In [25]:
# classic version (c or c++)
li = ["python","programming"]
print('%s %s ' % (li[0],li[1]))

python programming 


In [26]:
li = [1,2,3,4]
print('%d %d %d %d' % (li[0],li[1],li[2],li[3]))

1 2 3 4


In [27]:
# .format
li = ["python","programming"]
print("List Item 0 : {0} \nList Item 1: {1}".format(li[0],li[1]))

List Item 0 : python 
List Item 1: programming


In [28]:
li = [1,2,3,4]
print('{0} {1} {2} {3}'.format(li[0],li[1],li[2],li[3]))

1 2 3 4


### Data Structures
- List 
- Tuples
- Dictionaries

### Dictionaries
- It works on concept of unique data
- Key,value is the unique identifier a value
- Each key is seperated from value colon(:)
- Each key and value is seperated by comma(,)
- Dictionaries are enclosed with curly brackets({})


In [30]:
d1 = {"Name":"SANKAR","Email":"bhavanisankar5758@gmail.com","Address":"RAJUPALEM"}
print(d1)

{'Name': 'SANKAR', 'Email': 'bhavanisankar5758@gmail.com', 'Address': 'RAJUPALEM'}


In [31]:
d1["Email"] # Access the specific key value

'bhavanisankar5758@gmail.com'

In [33]:
d1["Email"] = "121822502018@gitam.in" # update the value using key

In [36]:
d1.keys() # Allthe keys of dict object

dict_keys(['Name', 'Email', 'Address'])

In [35]:
d1.values() # All the vale of value obj

dict_values(['SANKAR', '121822502018@gitam.in', 'RAJUPALEM'])

In [38]:
del d1["Email"]

In [37]:
d1

{'Name': 'SANKAR', 'Email': '121822502018@gitam.in', 'Address': 'RAJUPALEM'}

In [ ]:
d1

### contact Application
- Add contact details
- Search the contact details
- List all the contacts(merge/Import)
  - Name1:value
  - Name2:value
- Modify or update the contact details
- Delete the contact details

In [43]:
contact = {} # Empty Dict obj
# Let's create a function to add contact details
def addContact(name,phone):
    if name not in contact: # cond to check is exist or not
        contact[name] = phone
        print('contact details are added successfully')
    else:
        print('Name is already exists')
    return
addContact('Sankar','9000700878')
addContact('surendra','9885508424')
addContact('Narendra','9885166364')

contact details are added successfully
contact details are added successfully
contact details are added successfully


In [44]:
contact

{'Sankar': '9000700878', 'surendra': '9885508424', 'Narendra': '9885166364'}

In [47]:
def searchContact(name):
    if name in contact:
        print(name, " : ",contact[name])
    else:
        print('{0} Does not present'.format(name))
    return
searchContact('Sankar')
searchContact('Gitam')
searchContact('Narendra')


Sankar  :  9000700878
Gitam Does not present
Narendra  :  9885166364


In [58]:
# Import new contact details
# Merge the previous details with new details
def importContact(newContact):
    contact.update(newContact)
    print(len(newContact.keys()),'Contact are added')
    return
newContact = {'Gitam':123456767,'Anil':9885513133,'python':9676522793}
importContact(newContact)


3 Contact are added


In [59]:
contact

{'Sankar': '9000700878',
 'surendra': '9885508424',
 'Narendra': '9885166364',
 'Gitam': 123456767,
 'Anil': 9885513133,
 'python': 9676522793}

In [60]:
def modifyContact(name,phone):
    if name in contact:
        contact[name] = phone
        print(name,'Updated successfully')
    else:
        print(name,'Not exist')
    return
modifyContact('Gitam',9885175726)
modifyContact('Ravi',6500218756)

Gitam Updated successfully
Ravi Not exist


In [63]:
def deleteContact(name):
    if name in contact:
        del contact[name]
        print(name,'Deleted successfully')
    else:
        print(name,'Not Present')
    return
deleteContact('Anil')

Anil Deleted successfully


In [64]:
contact

{'Sankar': '9000700878',
 'surendra': '9885508424',
 'Narendra': '9885166364',
 'Gitam': 9885175726,
 'python': 9676522793}

### Tuples
- Tuples enclosed with parenthesis()

### Differences between Tuples and List
    - List are mutable -- can be changed / modify
      - Read and write the data
    - Tuples are immutable -- can't be changed / modify
      -Read data only

In [66]:
t1 = (1,2,3,4,5)
print(t1)

(1, 2, 3, 4, 5)


### Standard Libraries
- Regular Expression
  - To Validate the Indian Mobile
  - To Validate the EmailID
  - To Validate the username
  - To validate the password
- Understanding the Regular Expressions
  - [0-9] -- Any digit matching
  - [a-z] -- Any character matching

In [67]:
# Function to test the two digit
# True -- Input two digit number
# False -- Input not a two digit number
import re
def twoDigitMatching(n):
    pattern = '^[0-9]{2}$'
    n = str(n)    # str() converts input into str format
    if re.match(pattern,n):
        return True
    return False
print(twoDigitMatching(12))  # True
print(twoDigitMatching(123)) # False


True
False


In [71]:
# REgular Expression for Username
# Username -- should contains Lower and upper case characters
# Length  -- Min 5 Max 12
def validateUsername(s):
    pattern = '^[a-zA-Z]{4,11}$'
    if re.match(pattern,s):
        return True
    return False
print(validateUsername('sankarreddy'))
print(validateUsername('sankar'))
print(validateUsername('Bhavanisankarreddy'))

True
True
False


### Regular Expression 
- Regular Expression for RollNumber
   - Example : 121822502018 # valid
   - Example : 131925454683 # Not valid
- Regular Expression for password
   - Parameters - Len min

### Regular Expression for indian mobile
- 10 Digit Number
  - First number is starts from[6-9]
  - Rest of 9 numbers is between [0-9]
     - example : 9000700878 # True
     - example : 6515596046 # False
- 11 Digit Number
     - First number can be 0
     - Second number can be [6-9]
     - Rest of 9 numbers is between [0-9]
         - Example :09000700878
         - Example :99000700878
- 13 Digit number
   - First number can be +
   - Second number can be 9
   - Third number can be 1
    -fourth number [6-9]
   

In [ ]:
# Rule 1 :  ^[6-9][0-9]{9}$
# Rule 2 :  ^[0][6-9][0-9]{9}$
# Rule 3 :  ^[+][9][1][6-9][0-9]{9}$

In [1]:
import re
def validateIndiaPhone(n):
    pattern = '^[6-9][0-9]{9}$|^[0][6-9][0-9]{9}$|^[+][9][1][6-9][0-9]{9}$'
    n = str(n)
    if re.match(pattern,n):
        return True
    return False
validateIndiaPhone('+919000700878') 

True

### Validate the EmailID
- Example : Username@domainName.extension
- Example : dsps@gmail.com
- Username : 
   - Len of string will be [6-15]
   - No spl characters except underscore(_)
   - Username should not begins with underscore(_)
   - All characters set digits,Lower case,Upper case
      - Example : Sankar_5765
- DomainName
   - Len of string will be [3-18]
   - No spl characters
   - Characters set will be Digits,Lower case,Upper case
          - Example : gmail,hotmail,ymail,rediff and etc..
- Extension
  - Len of string will be [2-4]
  - No spl characters allowed
  -  Character set will be Lower case characters
     - Example : com in au uk
  



In [ ]:
# User name RE : [0-9a-zA-Z][0-9a-zA-Z_.]{5,14}
# Domain Name Re : [a-z0-9]{3,18}
# Extension RE : [a-z]{2,4}

In [3]:
def validateEmail(email):
    pattern = '^[0-9a-zA-Z][0-9a-zA-z_.]{5,14}[@][a-z0-9]{3,18}[.][a-z]{2,4}$'
    if re.match(pattern,email):
        return True
    return False 
validateEmail('bhavanisankar5758@gmailmail.com')

True

### Package Graphics
- Lib - turtle

### Package,Module and Sub Package
### Package 
   - A collection of modules(single python file .py)
### Module
   - A single python File contains group of functions
   
   - package - > sub package -> modules - > Functions - > Statements
   
   - import keyword used to import the external Lib from the python program

In [4]:
import math    # Import all the function into python program
math.floor(123.456)

123

In [5]:
math.ceil(123.456)

124

In [6]:
from math import factorial as fact  # will import only specific function
# as fact alias name
fact(10)

3628800

In [7]:
from math import gcd as gcd
gcd(10,15)

5

In [11]:
# Import random package
import random 
#  Function to generate the N Numbers in given range
def generateRandNum(n,lb,ub):
    for i in range(0,n):
        print(random.randint(lb,ub),end=' ')
    return
generateRandNum(10,0,100)

48 95 31 66 18 68 25 75 57 42 

In [ ]:
# Create a simple game
# Try to generate 20 numbers
# Input : Number
# Present : Congrats!!!
# Not present : Try again

In [12]:
from random import randint as rdint
def generateNumber(n,lb,ub):
    li = []
    for i in range (0,n):
        li.append(rdint(lb,ub))
    return li
generateNumber(20,100,999)
        
    

[142,
 860,
 588,
 960,
 529,
 615,
 948,
 255,
 384,
 341,
 926,
 279,
 737,
 394,
 344,
 528,
 201,
 432,
 587,
 379]

In [3]:
# Create a simple game
# Try to generate 20 numbers
# Input : Number
# Present : Congrats!!!
# Not present : Try again
from random import randint as rdint
def generateNumber(n,lb,ub):
    li = []
    for i in range (0,n):
        li.append(rdint(lb,ub))
    return li
def check(n):
    li = generateNumber(20,100,999)
    if n in li:
        print('Congrats!!!')
    else:
        print('Try again!!!')
    return
check(56)

Try again!!!
